mender-update/deployments_test.cpp

```diff
@@ -659,3 +659,225 @@ TEST_F(DeploymentsTests, TestV1APIFallbackWithError) {
+			EXPECT_EQ(req->GetMethod(), http::Method::PUT);
+			EXPECT_EQ(common::StringFromByteVector(received_body), expected_request_data);
+
+			auto result = req->MakeResponse();
+			ASSERT_TRUE(result);
+			auto resp = result.value();
+
+			resp->SetHeader("Content-Length", to_string(response_data.size()));
+			resp->SetBodyReader(make_shared<io::StringReader>(response_data));
+			resp->SetStatusCodeAndMessage(200, "Success");
+			resp->AsyncReply([](error::Error err) { ASSERT_EQ(error::NoError, err); });
+		});
+
+	bool handler_called = false;
+	auto err = deps::PushStatus(
+		deployment_id,
+		status,
+		substatus,
+		"http://127.0.0.1:" TEST_PORT,
+		client,
+		loop,
+		[&handler_called, &loop](deps::StatusAPIResponse resp) {
+			handler_called = true;
+			EXPECT_EQ(resp, error::NoError);
+			loop.Stop();
+		});
+	EXPECT_EQ(err, error::NoError);
+
+	loop.Run();
+	EXPECT_TRUE(handler_called);
+}
+
+TEST_F(DeploymentsTests, PushStatusFailureTest) {
+	TestEventLoop loop;
+
+	http::ServerConfig server_config;
+	http::Server server(server_config, loop);
+
+	http::ClientConfig client_config;
+	http::Client client {client_config, loop};
+
+	auto status {deps::DeploymentStatus::Installing};
+	string deployment_id = "2";
+	string substatus = "Installing now";
+	string expected_request_data = R"({"status":"installing","substate":")" + substatus + "\"}";
+
+	const string response_data = R"({"error": "Access denied", "response-id": "some id here"})";
+
+	vector<uint8_t> received_body;
+	server.AsyncServeUrl(
+		"http://127.0.0.1:" TEST_PORT,
+		[&received_body, &expected_request_data](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+			auto req = exp_req.value();
+
+			auto content_length = req->GetHeader("Content-Length");
+			ASSERT_TRUE(content_length);
+			EXPECT_EQ(content_length.value(), to_string(expected_request_data.size()));
+			auto ex_len = common::StringToLongLong(content_length.value());
+			ASSERT_TRUE(ex_len);
+
+			auto body_writer = make_shared<io::ByteWriter>(received_body);
+			received_body.resize(ex_len.value());
+			req->SetBodyWriter(body_writer);
+		},
+		[&received_body, &expected_request_data, &response_data, deployment_id](
+			http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+
+			auto req = exp_req.value();
+			EXPECT_EQ(
+				req->GetPath(),
+				"/api/devices/v1/deployments/device/deployments/" + deployment_id + "/status");
+			EXPECT_EQ(req->GetMethod(), http::Method::PUT);
+			EXPECT_EQ(common::StringFromByteVector(received_body), expected_request_data);
+
+			auto result = req->MakeResponse();
+			ASSERT_TRUE(result);
+			auto resp = result.value();
+
+			resp->SetHeader("Content-Length", to_string(response_data.size()));
+			resp->SetBodyReader(make_shared<io::StringReader>(response_data));
+			resp->SetStatusCodeAndMessage(403, "Unauthorized");
```

Comment:
403 is actually "Forbidden". Not that it matters here, but maybe change the text just to avoid confusion with the real "401 Unauthorized"? Which would trigger a different code path in production.

Reply:
Yeah, doesn't matter here, but better not seed confusion.